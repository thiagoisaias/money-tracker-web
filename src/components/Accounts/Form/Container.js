import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import Form from "./Form";

import {
  createAccount,
  updateAccount,
  setAccountToEdit,
  clearAccountsError,
  clearAccountToEdit
} from "store/actions/accounts/accounts";

export class Container extends Component {
  onSubmitData = formData => {
    const {
      accountToEdit,
      history,
      isLoading,
      match,
      onCreateAccount,
      onUpdateAccount
    } = this.props;

    if (isLoading) {
      return;
    }

    if (match.path === "/accounts/new") {
      onCreateAccount(formData, history);
    } else if (match.path === "/accounts/:id/edit" && accountToEdit) {
      onUpdateAccount(formData, accountToEdit.id, history);
    } else {
      return;
    }
  };

  render() {
    const { accountToEdit, error, isLoading, match } = this.props;
    return (
      <Form
        {...accountToEdit}
        onSubmitData={this.onSubmitData}
        error={error}
        isLoading={isLoading}
        match={match}
      />
    );
  }

  /* When the page is reloaded and the path is /account/:id/edit, accountToEdit is null, so the
  app redirects the user back to /accounts
  */
  componentDidMount() {
    const { accountToEdit, history, match } = this.props;
    if (match.path === "/accounts/:id/edit" && accountToEdit === null) {
      history.push("/accounts");
    }
  }

  // The value of accountToEdit should be null when the user leave /accounts/:id/edit
  componentWillUnmount() {
    const {
      error,
      match,
      onClearAccountsError,
      onClearAccountToEdit
    } = this.props;

    if (match.path === "/accounts/:id/edit") {
      onClearAccountToEdit();
    }

    if (error) {
      onClearAccountsError();
    }
  }
}

Container.propTypes = {
  accountToEdit: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    initialBalance: PropTypes.string.isRequired
  }),
  error: PropTypes.string,
  history: PropTypes.object.isRequired,
  isLoading: PropTypes.bool.isRequired,
  match: PropTypes.object.isRequired,
  onClearAccountToEdit: PropTypes.func.isRequired,
  onCreateAccount: PropTypes.func.isRequired,
  onUpdateAccount: PropTypes.func.isRequired,
  onSetAccountToEdit: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  accountToEdit: state.accounts.accountToEdit,
  isLoading: state.accounts.isLoading,
  error: state.accounts.error
});

const mapDispatchToProps = {
  onCreateAccount: createAccount,
  onSetAccountToEdit: setAccountToEdit,
  onClearAccountsError: clearAccountsError,
  onClearAccountToEdit: clearAccountToEdit,
  onUpdateAccount: updateAccount
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(Container)
);
