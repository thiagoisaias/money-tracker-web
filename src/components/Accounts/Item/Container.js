import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import { accountType } from "types";

import Item from "./Item";

import {
  deleteAccount,
  setAccountToEdit
} from "store/actions/accounts/accounts";

export class Container extends Component {
  static propTypes = {
    accountData: accountType.isRequired,
    handleActiveItem: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
    isActive: PropTypes.bool.isRequired,
    isLoading: PropTypes.bool.isRequired,
    deleteAccount: PropTypes.func.isRequired,
    setAccountToEdit: PropTypes.func.isRequired
  };

  handleEdit = () => {
    const { accountData, history, setAccountToEdit } = this.props;
    setAccountToEdit(accountData);
    history.push(`/accounts/${accountData.id}/edit`);
  };

  handleDelete = () => {
    const { accountData, isLoading, deleteAccount } = this.props;
    if (isLoading) {
      return;
    }

    if (!window.confirm("Are you sure you want to delete this account?")) {
      return;
    }
    deleteAccount(accountData.id);
  };

  render() {
    const { accountData, handleActiveItem, isActive } = this.props;
    return (
      <Item
        accountData={accountData}
        handleActiveItem={handleActiveItem}
        handleDelete={this.handleDelete}
        handleEdit={this.handleEdit}
        isActive={isActive}
      />
    );
  }
}

const mapStateToProps = state => ({
  isLoading: state.accounts.isLoading
});

const mapDispatchToProps = {
  deleteAccount,
  setAccountToEdit
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(Container)
);
