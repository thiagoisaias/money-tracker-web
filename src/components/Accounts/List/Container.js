import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import { accountType } from "types";

import List from "./List";

import {
  fetchAccounts,
  clearAccountsError
} from "store/actions/accounts/accounts";

export class Container extends Component {
  static propTypes = {
    accountList: PropTypes.arrayOf(accountType).isRequired,
    error: PropTypes.string,
    isLoading: PropTypes.bool.isRequired,
    fetchAccounts: PropTypes.func.isRequired
  };

  componentDidMount() {
    const { fetchAccounts } = this.props;

    fetchAccounts();
  }

  render() {
    const { accountList, error, isLoading } = this.props;
    const propsToPass = {
      accountList,
      error,
      isLoading
    };
    return <List {...propsToPass} />;
  }

  componentWillUnmount() {
    const { error, clearAccountsError } = this.props;

    if (error) {
      clearAccountsError();
    }
  }
}

const mapStateToProps = state => ({
  isLoading: state.accounts.isLoading,
  error: state.accounts.error,
  accountList: state.accounts.accountList
});

const mapDispatchToProps = {
  fetchAccounts,
  clearAccountsError
};

export default connect(mapStateToProps, mapDispatchToProps)(Container);
