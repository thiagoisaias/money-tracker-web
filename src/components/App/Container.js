import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import App from "./App";

export class Container extends Component {
  static propTypes = {
    isAuthenticated: PropTypes.bool.isRequired
  };

  render() {
    return <App isAuthenticated={this.props.isAuthenticated} />;
  }
}

const mapStateToProps = state => ({
  isAuthenticated: state.auth.isAuthenticated
});

export default withRouter(connect(mapStateToProps)(Container));
