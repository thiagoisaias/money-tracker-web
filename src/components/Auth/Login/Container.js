import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import { clearAuthError, login } from "store/actions/auth/auth";
import Form from "./Form";

export class Container extends Component {
  onSubmitData = formData => {
    const { history, isLoading, onLogin } = this.props;

    if (isLoading) {
      return;
    }

    onLogin(formData, history);
  };

  render() {
    const { onLogin, isLoading, error } = this.props;
    return (
      <Form
        error={error}
        isLoading={isLoading}
        onLogin={onLogin}
        onSubmitData={this.onSubmitData}
      />
    );
  }

  componentWillUnmount() {
    const { error, onClearAuthError } = this.props;
    if (error) {
      onClearAuthError();
    }
  }
}

Container.propTypes = {
  error: PropTypes.string,
  onClearAuthError: PropTypes.func.isRequired,
  onLogin: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  history: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  isLoading: state.auth.isLoading,
  error: state.auth.error
});

const mapDispatchToProps = dispatch => ({
  onClearAuthError: () => {
    dispatch(clearAuthError());
  },

  onLogin: (loginData, history) => {
    dispatch(login(loginData, history));
  }
});

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(Container)
);
