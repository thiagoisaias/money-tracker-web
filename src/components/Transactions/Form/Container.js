import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import { accountType, categoryType, transactionType } from "types";

import {
  clearTransactionsError,
  clearTransactionToEdit,
  createTransaction,
  updateTransaction
} from "store/actions/transactions/transactions";

import { fetchAccounts } from "store/actions/accounts/accounts";
import { fetchCategories } from "store/actions/categories/categories";

import Form from "./Form";

export class Container extends Component {
  static propTypes = {
    accountList: PropTypes.arrayOf(accountType),
    accountListLoading: PropTypes.bool.isRequired,
    categoryList: PropTypes.arrayOf(categoryType),
    categoryListLoading: PropTypes.bool.isRequired,
    transactionToEdit: transactionType,
    error: PropTypes.string,
    history: PropTypes.object.isRequired,
    isLoading: PropTypes.bool.isRequired,
    match: PropTypes.object.isRequired,
    clearTransactionError: PropTypes.func.isRequired,
    clearTransactionToEdit: PropTypes.func,
    createTransaction: PropTypes.func.isRequired,
    fetchAccounts: PropTypes.func.isRequired,
    fetchCategories: PropTypes.func.isRequired,
    updateTransaction: PropTypes.func.isRequired
  };

  onSubmitData = formData => {
    const {
      history,
      isLoading,
      match,
      createTransaction,
      updateTransaction,
      transactionToEdit
    } = this.props;

    if (isLoading) {
      return;
    }

    const { accountId, ...transformedData } = formData;

    if (match.path === "/transactions/new") {
      createTransaction(transformedData, accountId, history);
    } else if (match.path === "/transactions/:id/edit" && transactionToEdit) {
      updateTransaction(formData, transactionToEdit.id, accountId, history);
    } else {
      return;
    }
  };

  handleSelectOptions() {
    const { accountList, categoryList } = this.props;

    const accountOptionList = accountList.map(item => {
      return { value: item.id, label: item.name };
    });

    const categoryOptionList = categoryList.map(item => {
      return { value: item.id, label: item.name };
    });

    return { accountOptionList, categoryOptionList };
  }

  componentDidMount() {
    const {
      accountList,
      categoryList,
      history,
      match,
      fetchAccounts,
      fetchCategories,
      transactionToEdit
    } = this.props;

    if (categoryList.length === 0) {
      fetchCategories();
    }

    if (accountList.length === 0) {
      fetchAccounts();
    }

    if (match.path === "/transactions/:id/edit" && transactionToEdit === null) {
      history.push("/");
    }
  }

  componentWillUnmount() {
    const {
      error,
      match,
      clearTransactionError,
      clearTransactionToEdit
    } = this.props;

    if (match.path === "/transactions/:id/edit") {
      clearTransactionToEdit();
    }

    if (error) {
      clearTransactionError();
    }
  }

  render() {
    const {
      accountListLoading,
      categoryListLoading,
      error,
      isLoading,
      match,
      transactionToEdit
    } = this.props;

    const {
      accountOptionList,
      categoryOptionList
    } = this.handleSelectOptions();

    return (
      <Form
        accountOptionList={accountOptionList}
        accountListLoading={accountListLoading}
        categoryOptionList={categoryOptionList}
        categoryListLoading={categoryListLoading}
        error={error}
        isLoading={isLoading}
        match={match}
        onSubmitData={this.onSubmitData}
        transactionToEdit={transactionToEdit}
      />
    );
  }
}

Container.propTypes = {};

const mapStateToProps = state => ({
  accountList: state.accounts.accountList,
  accountListLoading: state.accounts.isLoading,
  categoryList: state.categories.categoryList,
  categoryListLoading: state.categories.isLoading,
  error: state.transactions.error,
  isLoading: state.transactions.isLoading,
  transactionToEdit: state.transactions.transactionToEdit
});

const mapDispatchToProps = {
  clearTransactionsError,
  clearTransactionToEdit,
  createTransaction,
  fetchAccounts,
  fetchCategories,
  updateTransaction
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(Container)
);
