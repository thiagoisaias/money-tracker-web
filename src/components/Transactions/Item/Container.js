import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import Item from "./Item";

import { deleteTransaction } from "store/actions/transactions/transactions";

export class Container extends Component {
  handleDelete = () => {
    const { isLoading, onDeleteTransaction, transactionData } = this.props;

    if (isLoading) {
      return;
    }

    if (!window.confirm("Are you sure you want to delete this transaction?")) {
      return;
    }
    onDeleteTransaction(transactionData.id);
  };

  render() {
    return <Item {...this.props} handleDelete={this.handleDelete} />;
  }
}

Container.propTypes = {
  transactionData: PropTypes.shape({
    id: PropTypes.number.isRequired,
    account: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      initialBalance: PropTypes.number.isRequired
    }).isRequired,
    date: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    category: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      color: PropTypes.string.isRequired
    }).isRequired,
    value: PropTypes.number.isRequired,
    transactionType: PropTypes.string.isRequired
  }).isRequired,
  handleActiveItem: PropTypes.func.isRequired,
  isActive: PropTypes.bool.isRequired,
  isLoading: PropTypes.bool.isRequired,
  onDeleteTransaction: PropTypes.func.isRequired
};

const mapStateToProps = state => {
  return {
    isLoading: state.transactions.isLoading
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onDeleteTransaction: transactionId => {
      dispatch(deleteTransaction(transactionId));
    }
  };
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(Container)
);
