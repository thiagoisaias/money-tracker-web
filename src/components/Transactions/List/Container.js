import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import moment from "moment";

import List from "./List";
import withExpandableItem from "hoc/withExpandableItem/withExpandableItem";

import { fetchTransactions } from "store/actions/transactions/transactions";

export class Container extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedDate: this.props.selectedDate || moment()
    };
  }

  handleMonthChange = direction => {
    switch (direction) {
      case "next":
        this.handleNextMonth();
        break;
      case "previous":
        this.handlePreviousMonth();
        break;
      default:
        return;
    }
  };

  handleNextMonth = () => {
    this.setState(prevState => {
      return { selectedDate: prevState.selectedDate.clone().add(1, "month") };
    });
  };

  handlePreviousMonth = () => {
    this.setState(prevState => {
      return {
        selectedDate: prevState.selectedDate.clone().subtract(1, "month")
      };
    });
  };

  componentDidMount() {
    const { onFetchTransactionList } = this.props;
    onFetchTransactionList();
  }

  render() {
    const {
      activeItemId,
      error,
      isLoading,
      handleActiveItem,
      transactionList
    } = this.props;
    return (
      <List
        activeItemId={activeItemId}
        error={error}
        isLoading={isLoading}
        handleActiveItem={handleActiveItem}
        handleMonthChange={this.handleMonthChange}
        selectedDate={this.state.selectedDate}
        transactionList={transactionList}
      />
    );
  }
}

Container.propTypes = {
  activeItemId: PropTypes.number,
  error: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  handleActiveItem: PropTypes.func.isRequired,
  onFetchTransactionList: PropTypes.func.isRequired,
  selectedDate: PropTypes.object,
  transactionList: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      account: PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        initialBalance: PropTypes.number.isRequired
      }).isRequired,
      category: PropTypes.shape({
        id: PropTypes.number.isRequired,
        color: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired
      }).isRequired,
      date: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      transactionType: PropTypes.string.isRequired,
      value: PropTypes.number.isRequired
    })
  )
};

const mapStateToProps = state => {
  return {
    error: state.transactions.error,
    isLoading: state.transactions.isLoading,
    transactionList: state.transactions.transactionList
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onFetchTransactionList: () => {
      dispatch(fetchTransactions());
    }
  };
};

export default withExpandableItem(
  connect(mapStateToProps, mapDispatchToProps)(Container)
);
