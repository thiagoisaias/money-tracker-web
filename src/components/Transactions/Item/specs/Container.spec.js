import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { Container } from "../Container";

describe("Transaction Item Container", () => {
  const mockProps = {
    
  };
  const component = shallow(<Container {...mockProps} />);

  it("renders properly", () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
