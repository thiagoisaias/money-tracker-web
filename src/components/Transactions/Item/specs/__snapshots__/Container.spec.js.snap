// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction Item Container renders properly 1`] = `
<Item
  handleActiveItem={[Function]}
  handleDelete={[Function]}
  isActive={false}
  isLoading={false}
  onDeleteTransaction={[Function]}
  transactionData={
    Object {
      "account": Object {
        "id": 1,
        "initialBalance": 3242,
        "name": "Lorem",
      },
      "category": Object {
        "color": "salmon",
        "id": 1,
        "name": "Home",
      },
      "date": "05-29-2018",
      "description": "Lorem Ipsum",
      "id": 1,
      "transactionType": "expense",
      "value": 324.32,
    }
  }
/>
`;
